File: D-C/merge-sort.cpp
#include <stdio.h>
int number = 8;

int size;
int sorted[8];
int count = 0;

void merge(int a[], int m, int middle, int n) {
	int i = m;
	int j = middle + 1;
	int k = m;
	
	// 작은 순서대로 배열에 삽입
	while (i <= middle && j <= n) {
		if(a[i] <= a[j]) {
			sorted[k] = a[i];
			i++;
		}
		else {
			sorted[k] = a[j];
			j++;
		}
		k++;
	}
	
	// 남은 데이터도 삽입
	if(i > middle) {
		for(int t = j; t <= n; t++) {
			sorted[k] = a[t];
			k++;
		}
	} else {
		for(int t = i; t <= middle; t++) {
			sorted[k] = a[t];
			k++;
		}
	}
	
	// 정렬된 배열을 삽입
	for(int t = m; t <= n; t++) {
		a[t] = sorted[t];
	} 
}

void mergeSort(int a[], int m, int n) { // arr[8], 0, 7 
	// 이외의 경우는 크기가 1개인 경우

	printf("a = %d, m = %d, n = %d \n", a, m, n);
	if(m < n) {
		int middle = (m + n) / 2; // 3
		printf("middle : %d\n", middle);
		mergeSort(a, m, middle);
		mergeSort(a, middle + 1, n);
		merge(a, m, middle, n);
	} 
}

int main(void) {
	int array[number] = {7, 6, 5, 8, 3, 5, 9, 1};
	printf("array[%d] = %d // array = %d \n", number, array[number], array);
	mergeSort(array, 0, number - 1);
	for(int i = 0; i < number; i++) {
		printf("%d ", array[i]);
	}
}

File: python/algorithm/everything_algorithm/kind/sorting-algorithms/merge-sort.py
def merge_sort(arr):
    if len(arr) > 1:
        mid = len(arr) // 2
        L = arr[:mid]
        R = arr[mid:]

        # 왼쪽과 오른쪽 부분을 각각 정렬
        merge_sort(L)
        merge_sort(R)

        i = j = k = 0

        # 병합 과정
        while i < len(L) and j < len(R):
            if L[i] < R[j]:
                arr[k] = L[i]
                i += 1
            else:
                arr[k] = R[j]
                j += 1
            k += 1

        # 왼쪽 부분에 남은 요소들을 배열에 복사
        while i < len(L):
            arr[k] = L[i]
            i += 1
            k += 1

        # 오른쪽 부분에 남은 요소들을 배열에 복사
        while j < len(R):
            arr[k] = R[j]
            j += 1
            k += 1

    return arr

# 예시 배열
arr = [38, 27, 43, 3, 9, 82, 10]

# 병합 정렬 실행
sorted_arr = merge_sort(arr)
print("정렬된 배열:", sorted_arr)
